File: packages/print-service-client/src/fiscal-boleta-preview.ts
import type { PosSaleTicketCompany } from "./pos-sale-ticket";

/**
 * Boleta electrónica simulada (Set BE) → agente KaiPrinters (`type: "fiscal-boleta-preview"`).
 */

export const FISCAL_BOLETA_PREVIEW_PAYLOAD_VERSION = 1;

/** Ancho timbre PDF417: 100% del ticket (≈ {@link PRINT_FORMAT_PRESETS} `contentWidthMm` 70 en 80 mm). */
export const FISCAL_PDF417_USE_FULL_RECEIPT_WIDTH = true;

/** Factor de altura extra del timbre (preview SVG + raster ESC/POS). */
export const FISCAL_PDF417_HEIGHT_SCALE = 1.5;

/** Escala bwip-js PDF417 para preview HTML (módulos más grandes = más alto y legible). */
export function fiscalPdf417PreviewScale(format: "ticket_58mm" | "ticket_80mm"): number {
  return format === "ticket_58mm" ? 4 : 5;
}

export type FiscalBoletaPreviewEmisor = {
  rut: string | null;
  legalName: string | null;
  businessActivity: string | null;
  address: string | null;
  commune: string | null;
  city: string | null;
  resolutionNumber: string | null;
  resolutionDate: string | null;
};

/** RUT genérico SII cuando la venta no tiene cliente identificado (DTE sí lo lleva; ticket impreso no). */
export const GENERIC_BOLETA_RECEPTOR_RUT = "66666666-6";

export type FiscalBoletaPreviewReceptor = {
  rut: string;
  name: string;
};

/** Normaliza RUT para comparar con {@link GENERIC_BOLETA_RECEPTOR_RUT}. */
export function normalizeBoletaReceptorRut(rut: string): string {
  return rut.replace(/\./g, "").trim().toUpperCase();
}

/** Venta anónima / sin RUT válido → no mostrar bloque Receptor en ticket impreso. */
export function shouldShowReceptorOnFiscalBoletaTicket(receptor: {
  rut: string;
}): boolean {
  return normalizeBoletaReceptorRut(receptor.rut) !== GENERIC_BOLETA_RECEPTOR_RUT;
}

export type FiscalBoletaPreviewLine = {
  name: string;
  quantity: number;
  unitPriceWithIva: number;
  exempt?: boolean;
  unitMeasure?: string | null;
  lineTotal: number;
};

export type FiscalBoletaPreviewTotals = {
  mntNeto: number;
  mntExe: number;
  iva: number;
  mntTotal: number;
};

export type FiscalBoletaPreviewPayload = {
  version: number;
  caso: string;
  folio: number;
  issuedAt: string;
  tipoDte: 39;
  isSimulated: boolean;
  emisor: FiscalBoletaPreviewEmisor;
  company: PosSaleTicketCompany;
  receptor: FiscalBoletaPreviewReceptor;
  /** Si false, ESC/POS/HTML omiten bloque Receptor. Default: derivado de {@link shouldShowReceptorOnFiscalBoletaTicket}. */
  showReceptorOnTicket?: boolean;
  lines: FiscalBoletaPreviewLine[];
  totals: FiscalBoletaPreviewTotals;
  observation?: string | null;
  timbrePdf417Payload: string;
  operatorName?: string | null;
};

export type FiscalBoletaPreviewPrintExtras = {
  filename: string;
  documentType?: string;
  internalFolio?: string;
  sourceApp?: string;
  purpose?: string;
  format?: string;
};

File: kai-printers-desktop/src-tauri/src/fiscal_boleta_preview.rs
//! Boleta electrónica simulada (Set BE) → JSON (`fiscal-boleta-preview`).

use anyhow::{Context, Result};
use serde::Deserialize;
use std::path::PathBuf;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FiscalBoletaPreviewEmisor {
    pub rut: Option<String>,
    pub legal_name: Option<String>,
    pub business_activity: Option<String>,
    pub address: Option<String>,
    pub commune: Option<String>,
    pub city: Option<String>,
    pub resolution_number: Option<String>,
    pub resolution_date: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FiscalBoletaPreviewCompany {
    #[serde(default)]
    pub razon_social: String,
    pub nombre_fantasia: Option<String>,
    pub rut: Option<String>,
    pub business_activity: Option<String>,
    #[serde(default)]
    pub logo_base64: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FiscalBoletaPreviewReceptor {
    pub rut: String,
    pub name: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FiscalBoletaPreviewLine {
    pub name: String,
    pub quantity: f64,
    pub unit_price_with_iva: f64,
    #[serde(default)]
    pub exempt: bool,
    pub unit_measure: Option<String>,
    pub line_total: f64,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FiscalBoletaPreviewTotals {
    pub mnt_neto: f64,
    pub mnt_exe: f64,
    pub iva: f64,
    pub mnt_total: f64,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FiscalBoletaPreview {
    #[serde(default)]
    #[allow(dead_code)]
    pub version: i32,
    pub caso: String,
    pub folio: i64,
    pub issued_at: String,
    pub tipo_dte: i32,
    #[serde(default)]
    pub is_simulated: bool,
    pub emisor: FiscalBoletaPreviewEmisor,
    #[serde(default)]
    pub company: Option<FiscalBoletaPreviewCompany>,
    pub receptor: FiscalBoletaPreviewReceptor,
    /// Si false, no imprimir bloque Receptor (venta sin cliente). Ausente → inferir por RUT genérico.
    #[serde(default)]
    pub show_receptor_on_ticket: Option<bool>,
    #[serde(default)]
    pub lines: Vec<FiscalBoletaPreviewLine>,
    pub totals: FiscalBoletaPreviewTotals,
    pub observation: Option<String>,
    pub timbre_pdf417_payload: Option<String>,
    #[serde(default)]
    pub operator_name: Option<String>,
}

pub fn parse_fiscal_boleta_preview_from_value(value: &serde_json::Value) -> Result<FiscalBoletaPreview> {
    serde_json::from_value(value.clone()).context("parse fiscal-boleta-preview")
}

pub fn write_fiscal_boleta_preview_escpos_from_value(
    dir: &PathBuf,
    value: &serde_json::Value,
) -> Result<PathBuf> {
    crate::fiscal_boleta_preview_escpos::write_fiscal_boleta_preview_escpos_from_value(dir, value)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_fiscal_boleta_preview_json() {
        let v = serde_json::json!({
            "version": 1,
            "caso": "CASO-1",
            "folio": 42,
            "issuedAt": "2026-06-28",
            "tipoDte": 39,
            "isSimulated": true,
            "emisor": { "rut": "1-9", "legalName": "Test" },
            "receptor": { "rut": "66666666-6", "name": "Cliente" },
            "lines": [{ "name": "Item", "quantity": 1, "unitPriceWithIva": 1000, "lineTotal": 1000 }],
            "totals": { "mntNeto": 840, "mntExe": 0, "iva": 160, "mntTotal": 1000 },
            "timbrePdf417Payload": "<TED version=\"1.0\"><DD><RE>1-9</RE><TD>39</TD><F>42</F></DD></TED>"
        });
        let t = parse_fiscal_boleta_preview_from_value(&v).unwrap();
        assert_eq!(t.caso, "CASO-1");
        assert_eq!(t.folio, 42);
    }
}
